src/App.tsx

```diff
@@ -1,24 +1,225 @@
+  }, []);
+
+  useEffect(() => {
+    let timerId: NodeJS.Timeout;
+
+    if (error) {
+      timerId = setTimeout(() => {
+        setError(null);
+      }, 3000);
+    }
+
+    return () => clearTimeout(timerId);
+  }, [error]);
+
+  const todosCount = todos.length;
+  const areTodosVisible = todosCount > 0;
+
+  const visibleTodos = useMemo(() => (
+    getFilteredTodos(todos, statusFilter)
+  ), [todos, statusFilter]);
+
+  const completedTodos = useMemo(() => (
+    todos.filter((todo) => todo.completed)
+  ), [todos, statusFilter]);
+
+  const activeTodos = useMemo(() => (
+    todos.filter((todo) => !todo.completed)
+  ), [todos, statusFilter]);
+
+  const hasSomeActiveTodos = activeTodos.length > 0;
+  const hasSomeCompletedTodos = completedTodos.length > 0;
+
+  const changeFilterStatus = useCallback((filterStaus: StatusFilter) => {
+    setFilterStatus(filterStaus);
+  }, []);
+
+  const closeNotifications = useCallback(() => setError(null), []);
+
+  const changeNotification = (errorMessage: string) => {
+    setError(errorMessage);
+  };
+
+  const changeTitle = (newTitle: string) => {
+    setTitle(newTitle);
+  };
+
+  const addTodo = useCallback((todoTitle: string) => {
+    setIsLoading(true);
+
+    const newTodo = {
+      userId: USER_ID,
+      title: todoTitle,
+      completed: false,
+    };
+
+    setTempTodo({
+      id: 0,
+      ...newTodo,
+    });
+
+    postTodo(newTodo)
+      .then((todoFromResponse) => {
+        setTodos([
+          ...todos,
+          todoFromResponse,
+        ]);
+      })
+      .catch(() => {
+        setError('Unable to add a todo');
+      })
+      .finally(() => {
+        setIsLoading(false);
+        setTempTodo(null);
+        setTitle('');
+      });
+  }, [todos]);
+
+  const removeTodo = useCallback((todoId: number) => {
+    setLoadingTodoIds((currentTodoIds) => [
+      ...currentTodoIds,
+      todoId,
+    ]);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos((currentTodos) => (
+          currentTodos.filter((todo) => todo.id !== todoId)
+        ));
+      })
+      .catch(() => setError('Unable to delete a todo'))
+      .finally(() => {
+        setLoadingTodoIds((currentIds) => (
+          currentIds.filter((id) => id !== todoId)
+        ));
+      });
+  }, [todos]);
+
+  const clearCompletedTodos = useCallback(() => {
+    completedTodos.forEach(async (todo) => {
+      removeTodo(todo.id);
+    });
+  }, [completedTodos]);
+
+  const updateTodo = useCallback((
+    todoId: number,
+    updatedValue: string | boolean,
+  ) => {
+    setLoadingTodoIds((currentTodoIds) => [
+      ...currentTodoIds,
+      todoId,
+    ]);
+
+    const updatedTodo: Partial<Todo> = {};
+
+    switch (typeof updatedValue) {
+      case 'string':
+        updatedTodo.title = updatedValue;
+        break;
+
+      case 'boolean':
+        updatedTodo.completed = updatedValue;
+        break;
+
+      default:
```

Comment:
If you need to check only two conditions and not return the default value, you can use a short construction:
```
const updatedFieldName = typeof updatedValue === 'boolean' ? 'completed' : 'title';
updatedTodo[updatedFieldName] = updatedValue;
```